pkg/types/key_sets.go

```diff
@@ -0,0 +1,157 @@
+package types
+
+import (
+	"context"
+	"errors"
+	"fmt"
+
+	"github.com/kong/go-database-reconciler/pkg/crud"
+	"github.com/kong/go-database-reconciler/pkg/state"
+	"github.com/kong/go-kong/kong"
+)
+
+// keySetCRUD implements crud.Actions interface.
+type keySetCRUD struct {
+	client *kong.Client
+}
+
+func keySetFromStruct(arg crud.Event) *state.KeySet {
+	set, ok := arg.Obj.(*state.KeySet)
+	if !ok {
+		panic("unexpected type, expected *state.KeySet")
+	}
+	return set
+}
+
+// Create creates a KeySet in Kong.
+// The arg should be of type crud.Event, containing the KeySet to be created,
+// else the function will panic.
+// It returns a the created *state.KeySet.
+func (s *keySetCRUD) Create(ctx context.Context, arg ...crud.Arg) (crud.Arg, error) {
+	event := crud.EventFromArg(arg[0])
+	set := keySetFromStruct(event)
+	createdSet, err := s.client.KeySets.Create(ctx, &set.KeySet)
+	if err != nil {
+		return nil, err
+	}
+	return &state.KeySet{KeySet: *createdSet}, nil
+}
+
+// Delete deletes a KeySet in Kong.
+// The arg should be of type crud.Event, containing the KeySet to be deleted,
+// else the function will panic.
+// It returns a the deleted *state.KeySet.
+func (s *keySetCRUD) Delete(ctx context.Context, arg ...crud.Arg) (crud.Arg, error) {
+	event := crud.EventFromArg(arg[0])
+	set := keySetFromStruct(event)
+	err := s.client.KeySets.Delete(ctx, set.ID)
+	if err != nil {
+		return nil, err
+	}
+	return set, nil
+}
+
+// Update updates a KeySet in Kong.
+// The arg should be of type crud.Event, containing the KeySet to be updated,
+// else the function will panic.
+// It returns a the updated *state.KeySet.
+func (s *keySetCRUD) Update(ctx context.Context, arg ...crud.Arg) (crud.Arg, error) {
+	event := crud.EventFromArg(arg[0])
+	set := keySetFromStruct(event)
+	updatedSet, err := s.client.KeySets.Create(ctx, &set.KeySet)
+	if err != nil {
+		return nil, err
+	}
+	return &state.KeySet{KeySet: *updatedSet}, nil
+}
+
+type keySetDiffer struct {
+	kind                      crud.Kind
+	currentState, targetState *state.KongState
+}
+
+// Deletes generates a memdb CRUD DELETE event for KeySets
+// which is then consumed by the differ and used to gate Kong client calls.
+func (d *keySetDiffer) Deletes(handler func(crud.Event) error) error {
+	currentSetSets, err := d.currentState.KeySets.GetAll()
```

Comment:
nit: wouldn't currentKeySets be more clear?

Reply:
I will clear it up in next PR.